docs/rules/rules.tex

```diff
@@ -0,0 +1,1526 @@
+  $$
+
+\end{description}
+
+\section{Instantiating parameters}
+\label{sec:params}
+
+\subsection{Height}
+\label{sec:height}
+
+The height of a chain is defined as:
+
+$$
+\height{b₁,…,b_n} = n
+$$
+
+\subsection{Slot leader}
+\label{sec:slot-leader}
+
+A \emph{number generator} is a function $f ∈ ℕ ⟶ S ⟶ ℕ×S$ for some set
+$S$, such that for every $n$ and $s$, $(i,s') = f(n)(s) ⟹ 0 ⩽ i <
+n$. For security, we would need to assume that $f(n)$ has good
+randomness properties, but we can safely ignore this aspect for the
+purpose of this document.
+
+In the rest of the section, we assume given a number generator $f$, a
+seed $s_0∈S$, and a stake distribution $d$. Let $N$, the \emph{total
+  stake}, be $N=∑_{a∈\agentids} d(a)$ (which is defined since $d$ has
+finite support).
+
+Let us define the sequences $(s_n)_{n∈ℕ}$ and $(u_n)_{n>0}$ as
+\begin{itemize}
+\item $s_0$ is as given above
+\item $(u_{n+1}, s_{n+1}) = f(N)(s_n)$
+\end{itemize}
+
+We can then define the slot leader function as
+
+$$
+\slotleader{d}{s_0}{n} = \min_{l∈\agentids} \left( u_n < ∑_{a⩽l} d(a) \right)
+$$
+
+That is, laying the agents in order on a line, and letting each take
+$d(a)$ unit of length. We select the agent whose extent on the line
+contains $u_n$.
+
+\subsection{Number generator}
+\label{sec:number-generator}
+
+A \emph{binary number generator} is a function $g ∈ ℕ ⟶ S ⟶ ℕ×S$ for
+some set $S$, such that for every $n$ and $s$,
+$(i,s') = f(n)(s) ⟹ 0 ⩽ i < 2^{kn}$ for some $k$. A typical example is
+to have $k=8$ for a generator which produces a byte at a time.
+
+Given a binary number generator $g$ such that $g(N)$ is surjective on its first
+component, our goal is to construct a general number generator $f$ as in
+Section~\ref{sec:slot-leader}, such that $f(N)$ must also be surjective on its
+first component.\improvement{I think it's better to just put the surjective
+  requirement in the definition of (binary) random generators.}
+
+Let us fix $N∈ℕ$ and $s₀∈S$. And define $B = \min_{n∈ℕ}N < 2^{kn}$,
+and $(s_n)_{n∈ℕ}$ and $(u_n)_{n>0}$ as\improvement{I could define
+  generators once and for all and these could derive from them. This
+  seems to require a dependent product though. We can work around this
+  defining a family of generators instead.}
+\begin{itemize}
+\item $s_0$ is as given above
+\item $(u_{n+1}, s_{n+1}) = g(B)(s_n)$
+\end{itemize}
+
+Let $n₀ = \min_{n>0} \left(\rem{2^{kB}}{N} ⩽ u_n < 2^{kB}\right)$.
+
+We define
+$$
+f(N)(s₀) = (\rem{u_{n₀}}{N}, s_{n₀})
+$$
+
+Note that, by definition of the remainder,
+$\{ p∈ℕ | \rem{2^{kB}}{N} ⩽ u_n < 2^{kB} \}$ has $qN$ elements for
+some $q$. From which we can deduce that $f(N)$ is surjective on its
+first component (and actually: if the frequency of each $p$ in $u_n$
+is the same, then the same property also holds of $f$).
+
+\subsection{Chain quality}
+\label{sec:chain-quality}
+
+The chain qualitity function is defined as
+$$
+\quality{n} = 2n
+$$
```

Comment:
Maybe just add a comment that this is yet another usage of the overloaded term chain quality here?